### PR TITLE
Do not ivalidate the cache entirely on lock file change

### PR DESCRIPTION
**Pull request by dsame**
**Description:**

Add a function `cache-utils.ts:repoHasYarn3ManagedCache` returns `true` if there's at least one project with yarn3 managed cache

Yarn3 managed cache enables safe use of the previous cache `cache.restoreCache(cachePaths, primaryKey, [keyPrefix])`
because obsolete dependences are removed by Yarn not letting cache to grow endlessly

**Related issue:**
https://github.com/actions/setup-node/issues/325

The PR initially contained Node version in the key but it was removed due to
https://github.com/actions/setup-node/issues/325#issuecomment-1378444559

**Check list:**
- [ ] Mark if documentation changes are required.
- [ ] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

**Review by marko-zivic-93 — APPROVED** [review]
From the code standpoint it looks ok to me. I am not sure about the semantics :)

File: src/cache-restore.ts [hunk 1]
```diff
@@ -36,12 +39,13 @@ export const restoreCache = async (
     );
   }
 
-  const primaryKey = `node-cache-${platform}-${packageManager}-${fileHash}`;
+  const keyPrefix = `node-cache-${platform}-${packageManager}-${stepId}-${nodeVersion}`;
```
Comment (belgattitude) [inline src/cache-restore.ts:42]: Curious - Why stepId ?. In my understanding / experience the cache is totally shareable between steps.

**Review by nikolai-laevskii — APPROVED** [review]
Looks good to me

File: src/cache-utils.ts [hunk 2]
```diff
@@ -202,6 +221,58 @@ export const getCacheDirectories = async (
...
+): Promise<boolean> => {
+  if (packageManagerInfo.name !== 'yarn') return false;
+
+  const cacheDependencyPath = core.getInput('cache-dependency-path');
+
+  const yarnDirs = cacheDependencyPath
+    ? await getProjectDirectoriesFromCacheDependencyPath(cacheDependencyPath)
+    : [''];
+
+  for (const dir of yarnDirs.length === 0 ? [''] : yarnDirs) {
```
Comment (IvanZosimov) [inline src/cache-utils.ts:267]:
```suggestion
  for (const dir of !yarnDirs.length ? [''] : yarnDirs) {
```

File: src/cache-utils.ts [hunk 3]
```diff
@@ -110,6 +110,20 @@ export const getPackageManagerInfo = async (packageManager: string) => {
   }
 };
 
+/**
+ * getProjectDirectoriesFromCacheDependencyPath is called twice during `restoreCache`
+ *  - first through `getCacheDirectories`
+ *  - second from `repoHasYarn3ManagedCache`
+ *
+ *  it contains expensive IO operation and thus should be memoized
```
Comment (MaksimZhukov) [inline src/cache-utils.ts:118]: Is this a typo?
```suggestion
 *  it contains expensive IO operation and thus should be memorised
```

Comment (dsame) [reply src/cache-utils.ts:118]: No, it is by called by intention https://en.wikipedia.org/wiki/Memoization

Comment (nikolai-laevskii) [reply src/cache-utils.ts:118]: @MaksimZhukov nope https://en.wikipedia.org/wiki/Memoization

**Comment by nikolai-laevskii** [comment]
I'll keep my green light. Couldn't find something out of order, still looks good to me. But don't forget to rename branch from `yarn-perfomance` to `yarn-performance` before merging it into actions

**Review by marko-zivic-93 — APPROVED** [review]
LGTM!

File: src/cache-utils.ts [hunk 4]
```diff
@@ -202,6 +221,55 @@ export const getCacheDirectories = async (
...
+ *  - if local cache is not explicitly enabled (not yarn3), return false
+ *  - return true otherwise
+ */
+const isCacheManagedByYarn3 = async (directory: string): Promise<boolean> => {
+  const workDir = directory || process.env.GITHUB_WORKSPACE || '.';
+
+  // if .yarn/cache directory exists the cache is managed by version control system
+  const yarnCacheFile = path.join(workDir, '.yarn', 'cache');
+  if (fs.existsSync(yarnCacheFile) && fs.lstatSync(yarnCacheFile).isDirectory())
+    return Promise.resolve(false);
```
Comment (IvanZosimov) [inline src/cache-utils.ts:240]:
```suggestion
    return false;
```
I'd simplify it as , async function wraps result in Promise by default.

File: src/cache-utils.ts [hunk 5]
```diff
@@ -202,6 +221,55 @@ export const getCacheDirectories = async (
...
+
+  // NOTE: yarn1 returns 'undefined' with rc = 0
```
Comment (IvanZosimov) [inline src/cache-utils.ts:242]:
```suggestion
  // NOTE: yarn1 returns 'undefined' with return code = 0
```

Comment (IvanZosimov) [reply src/cache-utils.ts:242]: I'd expend `rc` to `return code` to make it more clear.

**Review by IvanZosimov — APPROVED** [review]

**Review by MaksimZhukov — APPROVED** [review]

File: src/cache-utils.ts [hunk 6]
```diff
@@ -202,6 +222,91 @@ export const getCacheDirectories = async (
...
+    '${enableGlobalCache[12]}
+    '${enableGlobalCache[13]}
+    '${enableGlobalCache[14]}
+    `
+  );
+  // only local cache is not managed by yarn
+  const managed = enableGlobalCache.trim() === 'false';
+  if (managed) {
+    core.debug(`"${workDir}" dependencies are managed by yarn 3 locally`);
+    return Promise.resolve(true);
```
Comment (dmitry-shibanov) [inline src/cache-utils.ts:281]: we can change `Promise.resolve(true) -> true`

**Review by dmitry-shibanov — APPROVED** [review]

**Comment by belgattitude** [comment]
Hey thanks ! a small question

Does it disable the node_modules folder caching when yarn3 zip files are enabled ?

**Review by dusan-trickovic — APPROVED** [review]

**Comment by dsame** [comment]
> Does it disable the node_modules folder caching when yarn3 zip files are enabled ?

Hello @belgattitude, the PR does not change the caching strategy in part of node_modules folder, but yarn3 [replaces node_modules with PnP](https://yarnpkg.com/features/pnp) that is supposed to be managed through VCS

**Comment by arcanis** [comment]
Hey - I'm trying to understand this change. Does it mean that the best thing a Yarn project running with this action can do is to NOT use the global cache (`enableGlobalCache: false`) AND to also NOT check-in the `.yarn/cache` folder?

In Yarn v4 we're changing the default to make `enableGlobalCache: true` the default - would it make sense to keep it to `true` when the install is running on a GitHub action?

**Comment by arcanis** [comment]
Ping @dsame ? We'll soon release the 4.0 stable; do we need to automatically set `enableGlobalCache: false` on GH Actions?

**Comment by belgattitude** [comment]
Personally I wouldn't.

The enableGlobalCache new default won't be enough to get the benefits (action cache keys are part of the recipe too). I would let the setup responsability to setup/node or any tool/action/script that help setting up caches.

Applying different defaults depending on env might also create confusion. 

I'd prefer a section in the yarn doc that focuses on a simple recipe for ci and caches (possibly with less emphasis on offline mode or a reference to it as an advanced use case.).  

That would help people implementing cache strategies (azure, vercel... ) to identify quickly what variables are important, what they do and apply a strategy within their "platform" possibilities.

Congrats for Yarn 4.0

**Comment by MaksimZhukov** [comment]
Hello @arcanis!
> I'm trying to understand this change. Does it mean that the best thing a Yarn project running with this action can do is to NOT use the global cache (enableGlobalCache: false) AND to also NOT check-in the .yarn/cache folder?

Yes, if the user uses the local cache and `.yarn/cache` folder is not in the repository, the action will automatically try to restore previous cache even if it is outdated in order to speed up the build.

> In Yarn v4 we're changing the default to make enableGlobalCache: true the default - would it make sense for us to keep it to false when the install is running on a GitHub action, and only there?

Thank you for contacting us and letting us know, we really appreciate it! In our opinion, it is not necessary to leave `enableGlobalCache` set to `false` for GitHub Actions, because this inconsistency can lead to confusion. Users can change this option if they want to use the local cache. We will reflect this behaviour in the action documentation to make it clear.

**Comment by belgattitude** [comment]
@arcanis for context, I realized there was a thread about the new setup node cache.  It starts from there https://github.com/actions/setup-node/issues/325#issuecomment-1535094446

**Pull request merged**